File: app/src/main/java/org/linphone/call/telecom/TelecomHelper.java
package org.linphone.call.telecom;

/*
TelecomHelper.java
Copyright (C) 2019 Belledonne Communications, Grenoble, France

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

import android.annotation.SuppressLint;
import android.annotation.TargetApi;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.graphics.drawable.Icon;
import android.net.Uri;
import android.os.Bundle;
import android.telecom.PhoneAccountHandle;
import android.telecom.TelecomManager;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;
import org.linphone.LinphoneManager;
import org.linphone.LinphoneService;
import org.linphone.R;
import org.linphone.contacts.ContactsManager;
import org.linphone.contacts.LinphoneContact;
import org.linphone.core.Address;
import org.linphone.core.Call;
import org.linphone.core.Core;
import org.linphone.core.CoreListenerStub;
import org.linphone.core.tools.Log;
import org.linphone.utils.LinphoneUtils;

@TargetApi(23)
public class TelecomHelper {
    public static final String LINPHONE_TELECOM_EXTRA_CONTACT_NAME = "CONTACT_NAME";
    public static final String LINPHONE_TELECOM_EXTRA_APP_NAME = "APP_NAME";
    public static final String LINPHONE_TELECOM_EXTRA_APP_ICON = "APP_ICON";

    private Context mContext;
    private TelecomManager mTelecomManager;
    private LinphoneTelecomAccount mLinphoneTelecomAccount;
    private PhoneAccountHandle mPhoneAccountHandle;
    private ExternalToLinphoneTelecomBroadcastReceiver mReceiver;
    private CoreListenerStub mListener;

    public TelecomHelper(Context context) {
        mContext = context;

        mTelecomManager = (TelecomManager) mContext.getSystemService(Context.TELECOM_SERVICE);

        mLinphoneTelecomAccount = new LinphoneTelecomAccount(mContext);
        mPhoneAccountHandle = mLinphoneTelecomAccount.getHandle();

        mReceiver = new ExternalToLinphoneTelecomBroadcastReceiver();

        mListener =
                new CoreListenerStub() {
                    @Override
                    public void onCallStateChanged(
                            Core core, Call call, Call.State state, String message) {
                        switch (state) {
                            case End:
                            case Released:
                                onCallTerminated(call.getCallLog().getCallId());
                                break;
                            case Connected:
                                if (call.getDir() == Call.Dir.Outgoing) {
                                    sendToCS(
                                            LinphoneConnectionService.EXT_TO_CS_ESTABLISHED,
                                            call.getCallLog().getCallId());
                                }
                        }
                    }
                };
        Core core = LinphoneManager.getCore();
        core.addListener(mListener);

        Log.i("[Telecom Manager] Created");
    }

    public void destroy() {
        Core core = LinphoneManager.getCore();
        core.removeListener(mListener);
        Log.i("[Telecom Manager] Destroyed");
    }

    public void refreshLinphoneTelecomAccount() {
        Log.i("[Telecom Manager] Refreshing account");
        mLinphoneTelecomAccount.refreshPhoneAccount();
    }

    public boolean isAccountEnabled() {
        return mLinphoneTelecomAccount.getPhoneAccount().isEnabled();
    }

    public void disable() {
        mLinphoneTelecomAccount.unregister();
    }

    public void onIncomingCallReceived(Call call) {
        if (call == null) return;
        Log.i("[Telecom Manager] Incoming call received");

        Bundle extras = prepareBundle(call);

        mTelecomManager.addNewIncomingCall(mPhoneAccountHandle, extras);
        if (call.getCore().getCallsNb() == 1) {
            registerCallScreenReceiver();
        }
    }

    @SuppressLint("MissingPermission")
    public void onOutgoingCallStarting(Call call) {
        if (call == null) return;
        Log.i("[Telecom Manager] Outgoing call started");

        Bundle extras = prepareBundle(call);

        // Send the call to LinphoneConnectionService, received by onCreateOutgoingConnection
        mTelecomManager.placeCall(Uri.parse(call.getRemoteAddress().asStringUriOnly()), extras);

        if (call.getCore().getCallsNb() == 1) {
            registerCallScreenReceiver();
        }
    }

    public void onCallTerminated(String callId) {
        Log.i("[Telecom Manager] Call terminated");
        sendToCS(LinphoneConnectionService.EXT_TO_CS_END_CALL, callId);

        Core core = LinphoneManager.getCore();
        if (core != null && core.getCallsNb() == 0) {
            unRegisterCallScreenReceiver();
        }
    }

    private Bundle prepareBundle(Call call) {
        Bundle extras = new Bundle();

        Address address = call.getRemoteAddress();
        LinphoneContact contact = ContactsManager.getInstance().findContactFromAddress(address);
        String contactName =
                contact == null
                        ? LinphoneUtils.getAddressDisplayName(address)
                        : contact.getFullName();

        String bundleKey;
        if (call.getDir() == Call.Dir.Outgoing) {
            bundleKey = TelecomManager.EXTRA_OUTGOING_CALL_EXTRAS;
            extras.putString(
                    TelecomManager.EXTRA_CALL_BACK_NUMBER,
                    call.getCallLog().getFromAddress().asStringUriOnly());
        } else {
            bundleKey = TelecomManager.EXTRA_INCOMING_CALL_EXTRAS;
            extras.putString(TelecomManager.EXTRA_INCOMING_CALL_ADDRESS, address.asStringUriOnly());
        }

        extras.putParcelable(TelecomManager.EXTRA_PHONE_ACCOUNT_HANDLE, mPhoneAccountHandle);
        extras.putString(
                LinphoneConnectionService.EXT_TO_CS_CALL_ID, call.getCallLog().getCallId());

        Bundle b = new Bundle();
        b.putString(LINPHONE_TELECOM_EXTRA_CONTACT_NAME, contactName);
        b.putString(LINPHONE_TELECOM_EXTRA_APP_NAME, mContext.getString(R.string.app_name));
        b.putParcelable(
                LINPHONE_TELECOM_EXTRA_APP_ICON,
                Icon.createWithResource(mContext, R.drawable.linphone_logo));
        extras.putBundle(bundleKey, b);

        return extras;
    }

    private void registerCallScreenReceiver() {
        LocalBroadcastManager.getInstance(LinphoneService.instance())
                .registerReceiver(
                        mReceiver, new IntentFilter(LinphoneConnectionService.CS_TO_EXT_BROADCAST));
        Log.i("[Telecom Manager] Registered call screen receiver");
    }

    private void unRegisterCallScreenReceiver() {
        LocalBroadcastManager.getInstance(LinphoneService.instance()).unregisterReceiver(mReceiver);
        Log.i("[Telecom Manager] Unregistered call screen receiver");
    }

    private void sendToCS(int action, String callId) {
        Log.i(
                "[Telecom Manager] Sending action "
                        + getActionNameFromValue(action)
                        + " for call id "
                        + callId);
        Intent intent = new Intent(LinphoneConnectionService.EXT_TO_CS_BROADCAST);
        intent.putExtra(LinphoneConnectionService.EXT_TO_CS_ACTION, action);

        intent.putExtra(LinphoneConnectionService.EXT_TO_CS_CALL_ID, callId);
        if (action == LinphoneConnectionService.EXT_TO_CS_HOLD_CALL) {
            intent.putExtra(LinphoneConnectionService.EXT_TO_CS_HOLD_STATE, true);
        }

        LocalBroadcastManager.getInstance(LinphoneService.instance()).sendBroadcast(intent);
    }

    public static String getActionNameFromValue(int action) {
        switch (action) {
            case LinphoneConnectionService.CS_TO_EXT_ANSWER:
                return "CS_TO_EXT_ANSWER";
            case LinphoneConnectionService.CS_TO_EXT_END:
                return "CS_TO_EXT_END";
            case LinphoneConnectionService.CS_TO_EXT_TERMINATE:
                return "CS_TO_EXT_TERMINATE";
            case LinphoneConnectionService.CS_TO_EXT_ABORT:
                return "CS_TO_EXT_ABORT";
            case LinphoneConnectionService.CS_TO_EXT_HOLD:
                return "CS_TO_EXT_HOLD";
            case LinphoneConnectionService.CS_TO_EXT_UNHOLD:
                return "CS_TO_EXT_UNHOLD";
        }
        return "Unknown: " + action;
    }
}

File: app/src/main/java/org/linphone/telecom/TelecomHelper.kt
/*
 * Copyright (c) 2010-2021 Belledonne Communications SARL.
 *
 * This file is part of linphone-android
 * (see https://www.linphone.org).
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package org.linphone.telecom

import android.annotation.SuppressLint
import android.annotation.TargetApi
import android.content.ComponentName
import android.content.Context
import android.graphics.drawable.Icon
import android.net.Uri
import android.os.Bundle
import android.telecom.PhoneAccount
import android.telecom.PhoneAccountHandle
import android.telecom.TelecomManager
import android.telecom.TelecomManager.*
import kotlin.Exception
import org.linphone.LinphoneApplication.Companion.coreContext
import org.linphone.R
import org.linphone.core.Call
import org.linphone.core.Core
import org.linphone.core.CoreListenerStub
import org.linphone.core.tools.Log
import org.linphone.utils.LinphoneUtils
import org.linphone.utils.PermissionHelper
import org.linphone.utils.SingletonHolder

@TargetApi(29)
class TelecomHelper private constructor(context: Context) {
    companion object : SingletonHolder<TelecomHelper, Context>(::TelecomHelper)

    private val telecomManager: TelecomManager = context.getSystemService(Context.TELECOM_SERVICE) as TelecomManager

    private var account: PhoneAccount = initPhoneAccount(context)

    val connections = arrayListOf<NativeCallWrapper>()

    private val listener: CoreListenerStub = object : CoreListenerStub() {
        override fun onCallStateChanged(
            core: Core,
            call: Call,
            state: Call.State?,
            message: String
        ) {
            Log.i("[Telecom Helper] Call state changed: ${call.state}")

            try {
                if (call.dir == Call.Dir.Incoming && call.state == Call.State.IncomingReceived) {
                    onIncomingCall(call)
                } else if (call.dir == Call.Dir.Outgoing && call.state == Call.State.OutgoingProgress) {
                    onOutgoingCall(call)
                }
            } catch (se: SecurityException) {
                Log.e("[Telecom Helper] Exception while trying to place call: $se")
            }
        }
    }

    init {
        coreContext.core.addListener(listener)
        Log.i("[Telecom Helper] Created")
    }

    fun destroy() {
        coreContext.core.removeListener(listener)
        Log.i("[Telecom Helper] Destroyed")
    }

    fun isIncomingCallPermitted(): Boolean {
        val incomingCallPermitted = telecomManager.isIncomingCallPermitted(account.accountHandle)
        Log.i("[Telecom Helper] Is incoming call permitted? $incomingCallPermitted")
        return incomingCallPermitted
    }

    @SuppressLint("MissingPermission")
    fun isInManagedCall(): Boolean {
        // Don't use telecomManager.isInCall as our own self-managed calls will be considered!
        val isInManagedCall = telecomManager.isInManagedCall
        Log.i("[Telecom Helper] Is in managed call? $isInManagedCall")
        return isInManagedCall
    }

    fun isAccountEnabled(): Boolean {
        val enabled = account.isEnabled
        Log.i("[Telecom Helper] Is account enabled ? $enabled")
        return enabled
    }

    @SuppressLint("MissingPermission")
    fun findExistingAccount(context: Context): PhoneAccount? {
        if (PermissionHelper.required(context).hasReadPhoneStateOrPhoneNumbersPermission()) {
            var account: PhoneAccount? = null
            val phoneAccountHandleList: List<PhoneAccountHandle> =
                telecomManager.selfManagedPhoneAccounts
            val connectionService = ComponentName(context, TelecomConnectionService::class.java)
            for (phoneAccountHandle in phoneAccountHandleList) {
                val phoneAccount: PhoneAccount = telecomManager.getPhoneAccount(phoneAccountHandle)
                if (phoneAccountHandle.componentName == connectionService) {
                    Log.i("[Telecom Helper] Found existing phone account: $phoneAccount")
                    account = phoneAccount
                    break
                }
            }
            if (account == null) {
                Log.w("[Telecom Helper] Existing phone account not found")
            }
            return account
        } else {
            Log.e("[Telecom Helper] Can't search for existing phone account, missing permission(s)")
        }
        return null
    }

    fun updateAccount(newAccount: PhoneAccount?) {
        if (newAccount != null) {
            Log.i("[Telecom Helper] Updating account object: $newAccount")
            account = newAccount
        }
    }

    fun removeAccount() {
        if (account.isEnabled) {
            Log.w("[Telecom Helper] Unregistering phone account handler from telecom manager")
            telecomManager.unregisterPhoneAccount(account.accountHandle)
        } else {
            Log.w("[Telecom Helper] Account wasn't enabled, skipping...")
        }
    }

    fun findConnectionForCallId(callId: String): NativeCallWrapper? {
        return connections.find { connection ->
            connection.callId == callId
        }
    }

    private fun initPhoneAccount(context: Context): PhoneAccount {
        val account: PhoneAccount? = findExistingAccount(context)
        if (account == null) {
            Log.i("[Telecom Helper] Phone account not found, let's create it")
            return createAccount(context)
        }
        return account
    }

    private fun createAccount(context: Context): PhoneAccount {
        val accountHandle = PhoneAccountHandle(
            ComponentName(context, TelecomConnectionService::class.java),
            context.packageName
        )
        // Take care that identity may be parsed, otherwise Android OS may crash during startup
        // and user will have to do a factory reset...
        val identity = coreContext.core.defaultAccount?.params?.identityAddress?.asStringUriOnly()
            ?: coreContext.core.createPrimaryContactParsed()?.asStringUriOnly()
            ?: "sip:linphone.android@sip.linphone.org"

        val address = Uri.parse(identity)
            ?: throw Exception("[Telecom Helper] Identity address for phone account is null!")
        val account = PhoneAccount.builder(accountHandle, context.getString(R.string.app_name))
            .setAddress(address)
            .setIcon(Icon.createWithResource(context, R.drawable.linphone_logo_tinted))
            .setCapabilities(PhoneAccount.CAPABILITY_SELF_MANAGED)
            .setHighlightColor(context.getColor(R.color.primary_color))
            .setShortDescription(context.getString(R.string.app_description))
            .setSupportedUriSchemes(listOf(PhoneAccount.SCHEME_SIP))
            .build()

        try {
            telecomManager.registerPhoneAccount(account)
            Log.i("[Telecom Helper] Phone account created: $account")
        } catch (uoe: UnsupportedOperationException) {
            Log.e("[Telecom Helper] Unsupported Operation Exception: $uoe")
        } catch (e: Exception) {
            Log.e("[Telecom Helper] Exception: $e")
        }
        return account
    }

    private fun onIncomingCall(call: Call) {
        Log.i("[Telecom Helper] Incoming call received from ${call.remoteAddress.asStringUriOnly()}")

        val extras = prepareBundle(call)
        telecomManager.addNewIncomingCall(
            account.accountHandle,
            Bundle().apply {
                putBundle(EXTRA_INCOMING_CALL_EXTRAS, extras)
                putParcelable(EXTRA_PHONE_ACCOUNT_HANDLE, account.accountHandle)
            }
        )
    }

    @SuppressLint("MissingPermission")
    private fun onOutgoingCall(call: Call) {
        Log.i("[Telecom Helper] Outgoing call started to ${call.remoteAddress.asStringUriOnly()}")

        val extras = prepareBundle(call)
        telecomManager.placeCall(
            Uri.parse(call.remoteAddress.asStringUriOnly()),
            Bundle().apply {
                putBundle(EXTRA_OUTGOING_CALL_EXTRAS, extras)
                putParcelable(EXTRA_PHONE_ACCOUNT_HANDLE, account.accountHandle)
            }
        )
    }

    private fun prepareBundle(call: Call): Bundle {
        val extras = Bundle()
        val address = call.remoteAddress

        if (call.dir == Call.Dir.Outgoing) {
            extras.putString(
                EXTRA_CALL_BACK_NUMBER,
                call.remoteAddress.asStringUriOnly()
            )
        } else {
            extras.putParcelable(EXTRA_INCOMING_CALL_ADDRESS, Uri.parse(address.asStringUriOnly()))
        }

        extras.putString("Call-ID", call.callLog.callId)

        val contact = coreContext.contactsManager.findContactByAddress(call.remoteAddress)
        val displayName = contact?.name ?: LinphoneUtils.getDisplayName(call.remoteAddress)
        extras.putString("DisplayName", displayName)

        return extras
    }
}
